libdrgn/python/type.c

```diff
@@ -1531,47 +1731,72 @@ static DrgnType *Program_compound_type(Program *self, PyObject *args,
 		return NULL;
 	}
 
-	PyObject *cached_members;
+	size_t num_templates;
+	PyObject *cached_members = NULL;
+	PyObject *cached_templates = NULL;
 	bool can_cache_members = true;
+	bool can_cache_templates = true;
```

Comment:
When we access some attribute of a type, we translate the C attribute to Python and cache it for subsequent uses. When a type is created in Python, we translate the Python arguments to C to build the type. As an optimization, we can sometimes cache the Python argument as an attribute directly when the type is created (that's what all the `_PyDict_SetItemId()` calls at the end of this function do). We can't cache the list of members/parameters/template parameters if they use a callback, because they each need their own thunk. That's what these are for, and you do it for template parameters (as well as a call to `_PyDict_SetItemId()`).